src/main/scala/monocly/impl/OpticImpl.scala

```diff
@@ -0,0 +1,259 @@
+        optic1.modifyA(a => optic2.modifyA(f)(a))(s)
+
+  def composeNonEmptyTraversal[S, T, A, B, C, D](optic1: NonEmptyTraversalImpl[S, T, A, B], optic2: NonEmptyTraversalImpl[A, B, C, D]): NonEmptyTraversalImpl[S, T, C, D] = 
+    new NonEmptyTraversalImpl:
+      override def nonEmptyModifyA[F[+_]: Apply](f: C => F[D])(s: S): F[T] =
+        optic1.nonEmptyModifyA(a => optic2.nonEmptyModifyA(f)(a))(s)
+
+  def composeOptional[S, T, A, B, C, D](optic1: OptionalImpl[S, T, A, B], optic2: OptionalImpl[A, B, C, D]): OptionalImpl[S, T, C, D] = 
+    new OptionalImpl:
+      override def getOrModify(s: S): Either[T, C] = 
+        optic1
+          .getOrModify(s)
+          .flatMap(a => optic2.getOrModify(a).left.map(optic1.replace(_)(s)))
+
+      override def replace(d: D): S => T = optic1.modify(optic2.replace(d))
+      override def getOption(s: S): Option[C] = optic1.getOption(s) flatMap optic2.getOption
+      override def modifyA[F[+_]: Applicative](f: C => F[D])(s: S): F[T] = optic1.modifyA(optic2.modifyA(f))(s)
+      override def modify(f: C => D): S => T = optic1.modify(optic2.modify(f))
+  end composeOptional
+
+  def composeLens[S, T, A, B, C, D](optic1: LensImpl[S, T, A, B], optic2: LensImpl[A, B, C, D]): LensImpl[S, T, C, D] = 
+    new LensImpl:
+      override def get(s: S): C = optic2.get(optic1.get(s))
+      override def replace(d: D): S => T = optic1.modify(optic2.replace(d))
+      override def modifyF[F[+_]: Functor](f: C => F[D])(s: S): F[T] = optic1.modifyF(optic2.modifyF(f))(s)
+      override def modify(f: C => D): S => T = optic1.modify(optic2.modify(f))
+
+  def composePrism[S, T, A, B, C, D](optic1: PrismImpl[S, T, A, B], optic2: PrismImpl[A, B, C, D]): PrismImpl[S, T, C, D] = 
+    new PrismImpl:
+      override def getOrModify(s: S): Either[T, C] = optic1.getOrModify(s)
+                                                  .flatMap(a => optic2.getOrModify(a).left.map(optic1.replace(_)(s)))
+      override def reverseGet(d: D): T = optic1.reverseGet(optic2.reverseGet(d))
+      override def getOption(s: S): Option[C] = optic1.getOption(s) flatMap optic2.getOption
+
+  def composeIso[S, T, A, B, C, D](optic1: IsoImpl[S, T, A, B], optic2: IsoImpl[A, B, C, D]): IsoImpl[S, T, C, D] = 
+    new IsoImpl: 
+      composeSelf =>
+
+      override def get(s: S): C = optic2.get(optic1.get(s))
+      override def reverseGet(d: D): T = optic1.reverseGet(optic2.reverseGet(d))
+      override def reverse: IsoImpl[D, C, T, S] =
+        new IsoImpl:
+          override def get(d: D): T = optic1.reverseGet(optic2.reverseGet(d))
+          override def reverseGet(s: S): C = optic2.get(optic1.get(s))
+          override def reverse: IsoImpl[S, T, C, D] = composeSelf
+  end composeIso
+
+end OpticImpl
+
+trait OpticImpl[-S, +T, +A, -B]:
+  def andThen[C, D](optic2: OpticImpl[A, B, C, D]): OpticImpl[S, T, C, D]
+
+  def foldMap[M: Monoid](f: A => M)(s: S): M = sys.error("This optic does not support `foldMap`")
```

Comment:
I wonder if it would be safer to throw exceptions in each implementation At the moment, it is easy to forget to override it wherever it makes sense.

Reply:
Performance wise, I wonder if it would be better if all methods were final. We'll probably need to bench that later.